tests/transactional/verify_undelete_snapshots.pm

```diff
@@ -0,0 +1,67 @@
+# SUSE's openQA tests
+#
+# Copyright © 2020 SUSE LLC
+#
+# Copying and distribution of this file, with or without modification,
+# are permitted in any medium without royalty provided the copyright
+# notice and this notice are preserved.  This file is offered as-is,
+# without any warranty.
+
+# Summary: Verify that essential snapsots cannot be deleted, see https://jira.suse.com/browse/SLE-3804
+# Maintainer: QA SLE Functional YaST <qa-sle-yast@suse.de>
+
+use strict;
+use warnings;
+use base "opensusebasetest";
+use testapi;
+use transactional qw(rpmver get_utt_packages trup_call);
+
+
+# Verify that essential snapshots cannot be deleted, see https://jira.suse.com/browse/SLE-3804
+sub run {
+    select_console 'root-console';
+
+    my @snapshots_before = (split /\n/, script_output('snapper list --disable-used-space'));
+    my $current_snapshot_before;
+    foreach (@snapshots_before) {
+        if ($_ =~ /(?<current_snapshot_before>^\d+)\*/) { $current_snapshot_before = $+{current_snapshot_before} }
+    }
+
+    # Create some snapshots with a transactional update snapshot in the middle.
+    record_info "Check undel", "Snapshot #1 - Check that essential snapshots cannot be deleted, 
+                                creates snapshots and installs a package";
+    assert_script_run("snapper create -d \"Disposable snapshot #1\"");
+    get_utt_packages;
+    trup_call "ptf install" . rpmver('security');
+    assert_script_run("snapper create -d \"Disposable snapshot #2\"");
+
+    my @snapshots_after_update = (split /\n/, script_output('snapper list --disable-used-space'));
+    my $next_snap_after_update;
+    my $current_snap_after_update;
+    foreach (@snapshots_after_update) {
+        if ($_ =~ /(?<current_snap>^\d+)\-/) { $current_snap_after_update = $+{current_snap} }
+        if ($_ =~ /(?<next_snap>^\d+)\+/)    { $next_snap_after_update    = $+{next_snap} }
+    }
+    die('Current snapshot was not marked with -') if (!length $current_snap_after_update);
+    die('New snapshot not marked with +')         if (!length $next_snap_after_update);
+    die unless ($current_snapshot_before == $current_snap_after_update);
+
+    my $last_snapshot        = $snapshots_after_update[-1];
+    my $last_snapshot_number = (split ' ', $last_snapshot)[0];
+    my @expected             = ('current system', 'currently mounted', 'next to be mounted');
+    #  The line below simulates a distracted admin trying to delete all snapshots...
+    my @delete_response = (split /\n/, script_output("snapper delete 0-$last_snapshot_number 2>&1"));
```

Comment:
you have all the information to compose your expected data (as you now have all snap numbers) put some \n at the end of each sentence and match regex with that, maybe you need modifier //s to treat as one line, so you can avoid all the complexity below of the loop.

Reply:
The last time I tried to treat a multiple command as one line and parse it with a regex to match multiple patterns  just resulted in time wasted, I ended up splitting it in an array (lines 38-44) and then loop into it. The regex worked on the paper (and in a regex tester) but not IRL, it only returned empty strings. Also regexes can be hard to read, I'm not sure it would make the code look simpler. I don't understand the idea of adding some \n, as they are already here, then I assume treating the output as one line means you ignore those end of lines anyway. I'll look if it seems to work, but otherwise, loop we will.